### ISSUE TITLE
Support for s3 files

### ISSUE THREAD

**Issue by terrykong** [google/sentencepiece#292]
Hi,

I was wondering if there are plans to support training a sentence piece model where the input is a comma separated list of s3 files or glob? This would be a nice-to-have feature for very large datasets that live on a kubernetes cluster with an s3 object store (for example). 

Thanks!

**Comment by taku910** [comment]
There are no plans to support special file storage including s3.  Instead of supporting individual file storage, it might be more useful to directly feed TextIOBase or Iterator object as follows.

````
response = s3.get_object(Bucket=bucket, Key=key)
data = TextIOWrapper(response['Body'])
spm.SentencePieceTrainer.train(stream=data, arg='--vocab_size=8000...')
````
This is just an example, not implemented.  Any comments/suggestions?

**Comment by terrykong** [comment]
Yea, I like your suggestion. I feel like this would help in situations where the training data is too large to fit in memory or on disk.

Out of curiosity, is there anything in how any of the sentencepiece methods are implemented that would prohibit learning in parallel, e.g.,

```
with concurrent.futures.ThreadPoolExecutor(max_workers=4) as executor:
    future_to_ground_truth = {executor.submit(spm.SentencePieceTrainer.train, stream=makes3Iterator(bucket), arg='--vocab_size=8000...'): bucket for bucket in buckets}
    for future in concurrent.futures.as_completed(future_to_ground_truth):
         pass

spm.SentencePieceTrainer.joinModels([getModelName(bucket) for bucket in buckets])
```

**Comment by taku910** [comment]
The training is performed in parallel in C++ layer (multithreading) when using --mode=unigram.
In addition, all training data is loaded in memory anyway,  so stream input just helps to load corpus directory from external file storage without making a copy of them locally.

`--input_sentence_size=size` will limit the number of sentences (lines) for training. The sentences are randomly sample from the entire input lines.

**Comment by terrykong** [comment]
Sorry for derailing the conversation. But going back to your suggestion.

> There are no plans to support special file storage including s3. Instead of supporting individual file storage, it might be more useful to directly feed TextIOBase or Iterator object as follows.
> 
> ```
> response = s3.get_object(Bucket=bucket, Key=key)
> data = TextIOWrapper(response['Body'])
> spm.SentencePieceTrainer.train(stream=data, arg='--vocab_size=8000...')
> ```
> This is just an example, not implemented. Any comments/suggestions?

Do you foresee this being worked in as a feature anytime soon?

**Comment by taku910** [comment]
Hi, I finally support the feature to feed training data with any iterable object. Thank you.

https://github.com/google/sentencepiece/tree/master/python#training-without-local-filesystem
